[boschshc] Long Polling crashes regularly every one or two days

## Expected Behavior
Items reflecting Bosch Smart Home Things' read channels shall be updated after every change.

## Current Behavior
After one or two days the items (e.g. temperature, humidity, open/close states) are not updated anymore (values are frozen to the last updated value). A view in the openhab.log reveals this exception:

```log
2023-11-17 05:58:17.372 [WARN ] [mmon.WrappedScheduledExecutorService] - Scheduled runnable ended with an exception: 
com.google.gson.JsonSyntaxException: java.lang.IllegalStateException: Expected BEGIN_OBJECT but was STRING at line 1 column 1 path $
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:270) ~[?:?]
	at com.google.gson.Gson.fromJson(Gson.java:1058) ~[?:?]
	at com.google.gson.Gson.fromJson(Gson.java:1016) ~[?:?]
	at com.google.gson.Gson.fromJson(Gson.java:959) ~[?:?]
	at com.google.gson.Gson.fromJson(Gson.java:927) ~[?:?]
	at org.openhab.binding.boschshc.internal.devices.bridge.LongPolling.onLongPollComplete(LongPolling.java:195) ~[?:?]
	at org.openhab.binding.boschshc.internal.devices.bridge.LongPolling$1.lambda$0(LongPolling.java:164) ~[?:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539) ~[?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
	at java.lang.Thread.run(Thread.java:833) ~[?:?]
Caused by: java.lang.IllegalStateException: Expected BEGIN_OBJECT but was STRING at line 1 column 1 path $
	at com.google.gson.stream.JsonReader.beginObject(JsonReader.java:395) ~[?:?]
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:259) ~[?:?]
	... 12 more
```

## Possible Solution
No idea. I could dig deeper into this problem but I don't know where to begin. Any advice is welcome.

## Steps to Reproduce (for Bugs)
Let it go, let it fail. I just restart the binding or openhab itself. It happens regularly after one or two days.

## Your Environment
<!-- Include as many relevant details about the environment you experienced the bug in -->
* Version used: openHAB 4.0.4
* OS: Raspberry OS Lite, openHAB running in a docker container

Hi, okay, so today is update day for me.

In the meantime I had another json hiccup yesterday evening with the latest controller software: 10.18.3249-29655

```
2023-12-18 19:22:56.128 [WARN ] [mmon.WrappedScheduledExecutorService] - Scheduled runnable ended with an exception: 
com.google.gson.JsonSyntaxException: java.lang.IllegalStateException: Expected BEGIN_OBJECT but was STRING at line 1 column 1 path $
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:270) ~[?:?]
	at com.google.gson.Gson.fromJson(Gson.java:1058) ~[?:?]
	at com.google.gson.Gson.fromJson(Gson.java:1016) ~[?:?]
	at com.google.gson.Gson.fromJson(Gson.java:959) ~[?:?]
	at com.google.gson.Gson.fromJson(Gson.java:927) ~[?:?]
	at org.openhab.binding.boschshc.internal.devices.bridge.LongPolling.onLongPollComplete(LongPolling.java:195) ~[?:?]
	at org.openhab.binding.boschshc.internal.devices.bridge.LongPolling$1.lambda$0(LongPolling.java:164) ~[?:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539) ~[?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635) ~[?:?]
	at java.lang.Thread.run(Thread.java:833) ~[?:?]
Caused by: java.lang.IllegalStateException: Expected BEGIN_OBJECT but was STRING at line 1 column 1 path $
	at com.google.gson.stream.JsonReader.beginObject(JsonReader.java:395) ~[?:?]
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:259) ~[?:?]
	... 12 more
```

I could cry ... I thought update via docker is easy ... sorry, this might not belong here ...

```
2023-12-19 10:01:10.765 [ERROR] [ternal.service.BootFeaturesInstaller] - Error installing boot feature repository mvn:org.openhab.distro/openhab-addons/4.1.0.M5/xml/features

java.lang.RuntimeException: Error resolving artifact org.openhab.distro:openhab-addons:xml:features:4.1.0.M5: [Could not find artifact org.openhab.distro:openhab-addons:xml:features:4.1.0.M5 in central (https://repo1.maven.org/maven2/)] : mvn:org.openhab.distro/openhab-addons/4.1.0.M5/xml/features
```

--- a/bundles/org.openhab.binding.boschshc/src/main/java/org/openhab/binding/boschshc/internal/devices/bridge/LongPolling.java
+++ b/bundles/org.openhab.binding.boschshc/src/main/java/org/openhab/binding/boschshc/internal/devices/bridge/LongPolling.java
@@ -34,6 +34,8 @@ import org.openhab.binding.boschshc.internal.serialization.GsonUtils;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import com.google.gson.JsonSyntaxException;
+
 /**
  * Handles the long polling to the Smart Home Controller.
  *
@@ -141,17 +143,20 @@ public class LongPolling {
     }
 
     /**
-     * Start long polling the home controller. Once a long poll resolves, a new one is started.
+     * Start long polling the home controller. Once a long poll resolves, a new one
+     * is started.
      */
     private void longPoll(BoschHttpClient httpClient, String subscriptionId) {
-        logger.debug("Sending long poll request");
 
         JsonRpcRequest requestContent = new JsonRpcRequest("2.0", "RE/longPoll", new String[] { subscriptionId, "20" });
         String url = httpClient.getBoschShcUrl("remote/json-rpc");
         Request longPollRequest = httpClient.createRequest(url, POST, requestContent);
 
-        // Long polling responds after 20 seconds with an empty response if no update has happened.
-        // 10 second threshold was added to not time out if response from controller takes a bit longer than 20 seconds.
+        logger.debug("Sending long poll request: {} with headers {}", longPollRequest, longPollRequest.getHeaders());
+
+        // Long polling responds after 20 seconds with an empty response if no update
+        // has happened. 10 second threshold was added to not time out if response
+        // from controller takes a bit longer than 20 seconds.
         longPollRequest.timeout(30, TimeUnit.SECONDS);
 
         this.request = longPollRequest;
@@ -159,8 +164,9 @@ public class LongPolling {
         longPollRequest.send(new BufferingResponseListener() {
             @Override
             public void onComplete(@Nullable Result result) {
-                // NOTE: This handler runs inside the HTTP thread, so we schedule the response handling in a new thread
-                // because the HTTP thread is terminated after the timeout expires.
+                // NOTE: This handler runs inside the HTTP thread, so we schedule the response
+                // handling in a new thread because the HTTP thread is terminated after the
+                // timeout expires.
                 scheduler.execute(() -> longPolling.onLongPollComplete(httpClient, subscriptionId, result,
                         this.getContentAsString()));
             }
@@ -188,10 +194,28 @@ public class LongPolling {
         if (failure != null) {
             handleLongPollFailure(subscriptionId, failure);
         } else {
-            logger.debug("Long poll response: {}", content);
+            handleLongPollResponse(httpClient, subscriptionId, content);
+        }
+    }
 
-            String nextSubscriptionId = subscriptionId;
+    /**
+     * Attempts to parse and process the long poll response content.
+     * <p>
+     * If the response cannot be parsed as {@link LongPollResult}, an attempt is made to parse a {@link LongPollError}.
+     * In case a {@link LongPollError} is present with the code <code>SUBSCRIPTION_INVALID</code>, a re-subscription is
+     * initiated.
+     * <p>
+     * If the response does not contain syntactically valid JSON, a new subscription is attempted with a delay of 15
+     * seconds.
+     * 
+     * @param httpClient HTTP client which received the response
+     * @param subscriptionId Id of subscription the response is for
+     * @param content Content of the response
+     */
+    private void handleLongPollResponse(BoschHttpClient httpClient, String subscriptionId, String content) {
+        logger.debug("Long poll response: {}", content);
 
+        try {
             LongPollResult longPollResult = GsonUtils.DEFAULT_GSON_INSTANCE.fromJson(content, LongPollResult.class);
             if (longPollResult != null && longPollResult.result != null) {
                 this.handleResult.accept(longPollResult);
@@ -212,10 +236,14 @@ public class LongPolling {
                     }
                 }
             }
-
-            // Execute next run
-            this.longPoll(httpClient, nextSubscriptionId);
+        } catch (JsonSyntaxException e) {
+            this.handleFailure.accept(
+                    new LongPollingFailedException("Could not deserialize long poll response: '" + content + "'", e));
+            return;
         }
+
+        // Execute next run
+        this.longPoll(httpClient, subscriptionId);
     }
 
     private void handleLongPollFailure(String subscriptionId, Throwable failure) {

--- a/bundles/org.openhab.binding.boschshc/src/main/java/org/openhab/binding/boschshc/internal/devices/bridge/LongPolling.java
+++ b/bundles/org.openhab.binding.boschshc/src/main/java/org/openhab/binding/boschshc/internal/devices/bridge/LongPolling.java
@@ -147,12 +147,11 @@ public class LongPolling {
      * is started.
      */
     private void longPoll(BoschHttpClient httpClient, String subscriptionId) {
+        logger.debug("Sending long poll request");
 
         JsonRpcRequest requestContent = new JsonRpcRequest("2.0", "RE/longPoll", new String[] { subscriptionId, "20" });
         String url = httpClient.getBoschShcUrl("remote/json-rpc");
         Request longPollRequest = httpClient.createRequest(url, POST, requestContent);
-
-        logger.debug("Sending long poll request: {} with headers {}", longPollRequest, longPollRequest.getHeaders());
 
         // Long polling responds after 20 seconds with an empty response if no update
         // has happened. 10 second threshold was added to not time out if response

--- a/bundles/org.openhab.binding.boschshc/src/test/java/org/openhab/binding/boschshc/internal/devices/bridge/LongPollingTest.java
+++ b/bundles/org.openhab.binding.boschshc/src/test/java/org/openhab/binding/boschshc/internal/devices/bridge/LongPollingTest.java
@@ -54,6 +54,7 @@ import org.openhab.binding.boschshc.internal.exceptions.BoschSHCException;
 import org.openhab.binding.boschshc.internal.exceptions.LongPollingFailedException;
 
 import com.google.gson.JsonObject;
+import com.google.gson.JsonSyntaxException;
 
 /**
  * Unit tests for {@link LongPolling}.
@@ -249,9 +250,8 @@ class LongPollingTest {
     }
 
     @Test
-    void startLongPolling_receiveScenario()
+    void startLongPollingReceiveScenario()
             throws InterruptedException, TimeoutException, ExecutionException, BoschSHCException {
-        // when(httpClient.getBoschSmartHomeUrl(anyString())).thenCallRealMethod();
         when(httpClient.getBoschShcUrl(anyString())).thenCallRealMethod();
 
         Request subscribeRequest = mock(Request.class);
@@ -362,6 +362,50 @@ class LongPollingTest {
 
         Result result = mock(Result.class);
         bufferingResponseListener.onComplete(result);
+    }
+
+    /**
+     * Tests a case in which the Smart Home Controller returns a HTML error response that is not parsable as JSON.
+     * <p>
+     * See <a href="https://github.com/openhab/openhab-addons/issues/15912">Issue 15912</a>
+     */
+    @Test
+    void startLongPollingInvalidLongPollResponse()
+            throws InterruptedException, TimeoutException, ExecutionException, BoschSHCException {
+        when(httpClient.getBoschShcUrl(anyString())).thenCallRealMethod();
+
+        Request subscribeRequest = mock(Request.class);
+        when(httpClient.createRequest(anyString(), same(HttpMethod.POST),
+                argThat((JsonRpcRequest r) -> "RE/subscribe".equals(r.method)))).thenReturn(subscribeRequest);
+        SubscribeResult subscribeResult = new SubscribeResult();
+        when(httpClient.sendRequest(any(), same(SubscribeResult.class), any(), any())).thenReturn(subscribeResult);
+
+        Request longPollRequest = mock(Request.class);
+        when(httpClient.createRequest(anyString(), same(HttpMethod.POST),
+                argThat((JsonRpcRequest r) -> "RE/longPoll".equals(r.method)))).thenReturn(longPollRequest);
+
+        fixture.start(httpClient);
+
+        ArgumentCaptor<CompleteListener> completeListener = ArgumentCaptor.forClass(CompleteListener.class);
+        verify(longPollRequest).send(completeListener.capture());
+
+        BufferingResponseListener bufferingResponseListener = (BufferingResponseListener) completeListener.getValue();
+
+        String longPollResultContent = "<HTML><HEAD><TITLE>400</TITLE></HEAD><BODY><H1>400 Unsupported HTTP Protocol Version: /remote/json-rpcHTTP/1.1</H1></BODY></HTML>";
+        Response response = mock(Response.class);
+        bufferingResponseListener.onContent(response,
+                ByteBuffer.wrap(longPollResultContent.getBytes(StandardCharsets.UTF_8)));
+
+        Result result = mock(Result.class);
+        bufferingResponseListener.onComplete(result);
+
+        ArgumentCaptor<Throwable> throwableCaptor = ArgumentCaptor.forClass(Throwable.class);
+        verify(failureHandler).accept(throwableCaptor.capture());
+        Throwable t = throwableCaptor.getValue();
+        assertEquals(
+                "Could not deserialize long poll response: '<HTML><HEAD><TITLE>400</TITLE></HEAD><BODY><H1>400 Unsupported HTTP Protocol Version: /remote/json-rpcHTTP/1.1</H1></BODY></HTML>'",
+                t.getMessage());
+        assertTrue(t.getCause() instanceof JsonSyntaxException);
     }
 
     @AfterEach

